Paper: Sacred

Qwlouse opened the pull request:
Authors: Klaus Greff, Aaron Klein, Martin Chovanec, Frank Hutter, Jürgen Schmidhuber

Abstract:
We present a toolchain for computational research consisting of Sacred and the two supporting tools Labwatch and Sacredboard.
Sacred is an open source python framework which aims provide basic infrastructure for running computational experiments and independent of the methods and libraries used
Instead it focusses on solving  universal everyday problems like managing configurations, reproducing results, and bookkeeping.
It also aims to provide an extensible basis for other tools, two of which we present here: Labwatch helps with tuning hyperparameters, and Sacredboard provides a web-dashboard for organizing and analyzing runs and results.

mpacer commented:
This paper does not build locally. I have verified, this is not a build server error.

jackiekazil commented:
Ha! I tried at the same time! (Delay in posting my comment, b/c of 2fa issues). Looking forward to seeing it built to provide feedback.

jjhelmus commented on papers/klaus_greff/paper.rst:
```diff
@@ -0,0 +1,759 @@
+
+
+Labwatch offers a simple and flexible interface to a variety of state-of-the-art hyperparameter optimization methods, including:
+- **Random search** is probably the simplest hyperparameter optimization method :cite:`bergstra-jmlr12a`. It just samples hyperparameter
```
A blank line is needed before this line to start the list

jjhelmus commented on papers/klaus_greff/paper.rst:
a blank line is needed after this comment to end the markup

jjhelmus commented on papers/klaus_greff/paper.rst:
This line should not be indented, either un-intent or join with the previous line

jjhelmus commented:
I added a few comments with suggested changes to get the paper to build.  @Qwlouse let me know if you need any help with these.

fchirigati commented:
Hi, my name is Fernando, and I've been asked to review this paper.

**Summary**

This paper presents Sacred, a framework that helps managing the execution of Python experiments. In a nutshell, it records the provenance of the runs (including configuration parameters, source code, package dependencies, and host information), keeps the "randomness" of the experiments (if any) for reproducibility purposes, and allows experiments to be monitored by observers (bookkeeping). It also comes with an assistant for selecting hyperparameter optimizers, and a user interface for browsing all the runs.

The motivation of the paper is very clear and the paper is very easy to follow and read. However, I see three main issues with the current version: (1) I'm missing a better comparison with related work (and some related work is also missing); (2) Some details of the framework are missing to clarify the contributions; (3) A discussion on the tool's limitations is missing. Find more details below.

**Main Issues**

(1) Managing experiment runs is not a new problem, and a better comparison against related work must be done:

- Sumatra is very similar to Sacred (also capturing package dependencies, having an interface to browse the different runs, etc.), and I did not grasp the real difference between them. The issues mentioned in the paper ("enforces a specific workflow [...] parameters need to be in a separate file [...] experiment must be an executable that takes the name of a config-file [...]") do not look like drawbacks to me. In the same way that Sacred is a bit intrusive, requiring users to use specific annotations, objects, and classes to run their experiments, Sumatra also has its requirements to make it possible to capture the provenance, etc.: both tools are in a way intrusive, requiring users to modify their code / workflow. The question is: why Sumatra's requirements are too bad compared to Sacred's? What is fundamentally different between them? Why would I use Sacred instead of Sumatra, or vice-and-versa? This needs to be better discussed.
- Scientific workflow systems -- such as [Taverna](http://www.taverna.org.uk/), [Kepler](https://kepler-project.org/), and [VisTrails](https://www.vistrails.org/index.php/Main_Page) (papers are listed on their main webpages) -- are well-known tools for capturing provenance, managing experiments, and helping design reproducible research. A comparison against them must be added to the related work.
- There are also other tools aimed at capturing the provenance of Python scripts that are missing from the related work. Perhaps the most relevant one is [noWorkflow](https://github.com/gems-uff/noworkflow), which also captures variables, source code / functions, module dependencies, etc. How are the techniques used in Sacred different from the ones used in noWorkflow? Other relevant tools include the one proposed [here](https://link.springer.com/chapter/10.1007/978-3-540-89965-5_24) and [ProvenanceCurious](http://dl.acm.org/citation.cfm?id=2452475).

A general suggestion for the authors is to take a look at papers published in provenance conferences / workshops (such as IPAW and TaPP), since managing experiments is a recurrent topic at these venues.

(2) There are some details missing that could clarify the contributions of the proposed framework. In general, I suggest authors to better explain and detail all the techniques used to capture the provenance of the runs. This also helps in comparing against existing work.

- Reproducibility Section, first paragraph: "Sacred [...] captures most of the relevant information." How does Sacred know what is relevant or not? And how this impacts reproducibility, i.e., in which cases there could be a false negative (relevant information that is not captured)?
- Reproducibility Section, second paragraph: "Relevant source-files are automatically detected through inspection". What is a relevant source file? And which kind of inspection is done? Any specific technique is implemented?
- Reproducibility Section, third paragraph: "Sacred detects imported packages and determines their version-numbers by inspection [...]  It might, however, miss certain nested imports" How are the packages and version numbers detected? I don't know what "inspection" means. Also, in which situations are nested imports missed? How deep does Sacred go into the dependency tree? Is there a way to specify how deep you go such as in noWorkflow?
- In terms of reproducibility, can you automatically re-run a previous experiment? Is there any user interface for that?
- How is the bookkeeping integrated to the user interface? I understand that the interface is connected to the database (MongoDB), but can one see the results being generated on the fly through the interface?

(3) While the authors do mention a few limitations under Future Work, a better discussion is needed. Some examples of discussion include (but are not limited to):

- What are the types of experiments that Sacred cannot support (if any)?
- In which cases the information captured by Sacred cannot assure reproducibility (e.g.: in which cases not all dependencies are captured)?
- What's the overhead in re-writing existing experiments to use Sacred?

**Nitpicking**

- There are missing commas in many sentences, e.g.: "Due to deadline pressure and the inherently unpredictable nature of research [,] there is ..."; "Through storage of run information in a central database [,] comprehensive query and sorting functionality for bookkeeping becomes available". Please double-check the paper for these cases.
- After colon, use lower case instead of upper case.
- Page 5, second column, first paragraph: " [...] a TensorFlow log has been created during the run, it is possible **_to_** launch TensorBoard directly from the Run detail view."
- Page 5, footnote 4: the text indicates that plots are already available, but the footnote contradicts this. Instead, simply say in the text that work is in progress.
- Future work: "Sacred has been useful for a many researchers [...]" Do you have real examples to share? Cases of success help show the contributions of the work!

cbcunc commented:
On page 2:

Thus in our example **log_filename** would be set to "log/NN64".

should be:

Thus in our example **log_dir** would be set to "log/NN64".

in order to match the example.

cbcunc commented:
On page 3, the statement:

Events are fired when a run is started, **every couple of seconds** during a run (heartbeat), and once it stops (either successfully or by failing).

is contradicted in the list of events which follow the statement:

Heartbeat Event
Fired continuously **every 10 seconds** while the experiment  is  running.  Contains  the  beat  time, captured stdout/stderr,  custom  information,  and  preliminary result.

cbcunc commented:
On page 6:

[Job12] is a Python library for scheduling lots of machine learning experiments which also helps in organizing hyperparameter searches  and  bookkeeping.

Should be:

Jobman [Job12] is a Python library for scheduling lots of machine learning experiments which also helps in organizing hyperparameter searches  and  bookkeeping.

cbcunc commented:
On page 6:

The most **wellknown** example might be the OpenML project [VvRBT14].

should be:

**The most well-known example might be the OpenML project [VvRBT14].**

cbcunc commented:
I'm not sure that this employ new or innovative methods or approaches to a problem.

Qwlouse commented:
Hi,

thanks for the constructive feedback. Especially to @jjhelmus for the pointers to related work. They were really helpful. 

We've fixed the typos and small mistakes, and I've significantly extended the treatment of related work. Tomorrow I will work on clarifying the requested details and on extending the discussion. 

> Do you have real examples to share? Cases of success help show the contributions of the work!

This is tricky. So far we got nothing that would qualify for a scientific text.
I know from personal communication and from the activity on the GitHub repository (which has accumulated >500 stars) that Sacred is being used and considered helpful. But that hardly seems like appropriate information for a paper.  And since there is not yet any publication about Sacred to cite, so I cannot yet point to published results that do.

> I'm not sure that this employs new or innovative methods or approaches to a problem.

While I mostly agree with this objection (Sacred is largely a combination of many known approaches), I'd like to disagree with the implied judgment. In my opinion, the significant contribution of Sacred is to bring the many solutions to all the everyday problems of computational research together into a single, unified, convenient system. Today, none of these small problems are unsolvable. But the combined effort required to solve them presents a big obstacle to reproducible data science. 
I'd also like to add that, to the best of my knowledge, Sacred is the only Python package that uses the principle of dependency injection, to simplify handling of configuration parameters.

fchirigati commented:
Hi @Qwlouse,

It was me who gave you the pointers to related work :)

Let me know when you have a complete revised version so I can take a look. When I first read the paper, I couldn't really grasp the difference between Sacred and other tools (e.g.: Sumatra) but hopefully you were able to make that clear. Detailing the techniques used by Sacred are also important to understand the contributions of the tool.

Talk to you soon!

Qwlouse commented:
Hi Fernando,
apologies for the confusion. Thanks to you of course for the excellent feedback! 
Also, I'm very sorry for the delay in incorporating your suggestions. I caught a fever on Friday and was unable to work for a few days. But since today I'm operational again and have rewritten the Reproducibility section to clarify many details and answer most of your questions. Regarding your 3rd point some of your questions were unclear to me: 

> What are the types of experiments that Sacred cannot support (if any)?

I'm not quite sure how to answer this. Support is restricted to Python code. Also, as outlined in the first paragraph of Future Work, there is no clear workflow for using Sacred with separate training and evaluation scripts, or when resuming aborted runs. But these points are mentioned already. Is there anything else that you feel is missing here? 

> What's the overhead in re-writing existing experiments to use Sacred?

I could discuss this point in more length, but it will heavily depend on how the code is structured already. The critical bit is usually how configuration values are handled. Other than that it is just about instantiating an Experiment and adding a decorator to the main function. Is something like this what you had in mind?

Other than that I think I have addressed all of your concerns. I know that there is little time left until the review deadline, but I hope you can still find the time to review my changes. 

Thanks again,
Klaus

katyhuff commented:
Thank you for your reviews, @fchirigati @jjhelmus @cbcunc ! The author seems to have made changes according to your comments. Do you feel that this paper is now ready for inclusion in the SciPy proceedings?

cbcunc commented:
@proceedings, @katyhuff, the paper is ready to be merged.

fchirigati commented:
Hi @Qwlouse ,

> I'm not quite sure how to answer this. Support is restricted to Python code. Also, as outlined in the first paragraph of Future Work, there is no clear workflow for using Sacred with separate training and evaluation scripts, or when resuming aborted runs. But these points are mentioned already. Is there anything else that you feel is missing here?

No, I think that's fine. When re-reading the paper, this was clearer to me, sorry for the confusion.

> I could discuss this point in more length, but it will heavily depend on how the code is structured already. The critical bit is usually how configuration values are handled. Other than that it is just about instantiating an Experiment and adding a decorator to the main function. Is something like this what you had in mind?

Yes, that's what I had in mind. I was just trying to understand what the burden in using Sacred would be for the user, but what you mentioned made sense.

Thanks for working on this revised version! You did address all my comments and I'm happy with the new version!

fchirigati commented:
@proceedings, @katyhuff The paper is ready to be merged.

Qwlouse commented:
Sorry, this is a bit late, but I wanted to thank the reviewers again for their constructive feedback. The quality of the paper has improved substantially, and I've learned about some related work that I had been missing. A special thank to @fchirigati for that!

fchirigati commented:
You are welcome, @Qwlouse ! The paper did improve substantially, and it is a very interesting (and important) subject! Congratulations on the work!
